File: src/ios/CDVBMSRequest.swift
/*
    Copyright 2016 IBM Corp.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
        http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

import Foundation
import BMSCore

@objc(CDVBMSRequest) class CDVBMSRequest : CDVPlugin {

    func send(_ command: CDVInvokedUrlCommand) {
        #if swift(>=3.0)
        self.commandDelegate!.run(inBackground: {
            let requestDict = command.arguments[0] as! NSDictionary
            var bodyData : Data? = nil

            let nativeRequest = self.unPackRequest(requestDict)

            if let body = requestDict.object(forKey: "body") as? String {
                bodyData = body.data(using: String.Encoding.utf8)
            }

            nativeRequest.send(requestBody: bodyData, completionHandler: { (response: Response?, error:Error?) in
                var responseString: String?
                do {

                    if (error != nil) {
                        // process the error
                        try responseString = self.packResponse(response,error: error)
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: responseString)
                        self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                    } else {
                        // process success
                        try responseString = self.packResponse(response)
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: responseString)
                        self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                    }
                } catch {
                    responseString = "Error Parsing JSON response."
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: responseString)
                    self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                }
            })

        })

        #else
            self.commandDelegate!.runInBackground({
                let requestDict = command.arguments[0] as! NSDictionary
                var bodyData : NSData? = nil

                let nativeRequest = self.unPackRequest(requestDict)

                if let body = requestDict.objectForKey("body") as? String {
                    bodyData = body.dataUsingEncoding(NSUTF8StringEncoding)
                }

                nativeRequest.send(requestBody: bodyData, completionHandler: { (response: Response?, error:NSError?) in
                    var responseString: String?
                    do {

                        if (error != nil) {
                            // process the error
                            try responseString = self.packResponse(response,error: error)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString:responseString)
                            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                        } else {
                            // process success
                            try responseString = self.packResponse(response)
                            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString: responseString)
                            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                        }
                    } catch {
                        responseString = "Error Parsing JSON response."
                        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: responseString)
                        self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                    }
                })

            })
        #endif
    }

    #if swift(>=3.0)
        fileprivate func unPackRequest(_ requestDict:NSDictionary) -> Request {
            // create a native request
            var url     = requestDict.object(forKey: "url") as! String

            // Detect if url is relative and convert to absolute
            if((url ?? "").isEmpty == false && url[url.startIndex] == "/") {
                url = convertRelativeURLToBluemixAbsolute(url)

            }

            // method
            let methodString = requestDict.object(forKey: "method") as? String
            var method : HttpMethod = HttpMethod.GET;
            switch(methodString){
                case "GET"?:
                    method = HttpMethod.GET
                    break
                case "POST"?:
                    method = HttpMethod.POST
                    break
                case "PUT"?:
                    method = HttpMethod.PUT
                    break
                case "DELETE"?:
                    method = HttpMethod.DELETE
                    break
                case "TRACE"?:
                    method = HttpMethod.TRACE
                    break
                case "HEAD"?:
                    method = HttpMethod.HEAD
                    break
                case "OPTIONS"?:
                    method = HttpMethod.OPTIONS
                    break
                case "CONNECT"?:
                    method = HttpMethod.CONNECT
                    break
                case "PATCH"?:
                    method = HttpMethod.PATCH
                    break
                default:
                    method = HttpMethod.GET
            }

            // get the query parameters
            let requestQueryParamsDict = requestDict.object(forKey: "queryParameters") as! Dictionary<String,String>

            // timeout
            let timeout = (requestDict.object(forKey: "timeout") as! Double) / Double(1000)

            // get the headers
            let requestHeaderDict = requestDict.object(forKey: "headers") as! Dictionary<String,String>

            let nativeRequest = Request(url: url, method: method, headers: requestHeaderDict, queryParameters: requestQueryParamsDict, timeout: timeout)
            return nativeRequest
        }
    #else
        private func unPackRequest(requestDict:NSDictionary) -> Request {
                // create a native request
                var url     = requestDict.objectForKey("url") as! String

                // Detect if url is relative and convert to absolute
                if((url ?? "").isEmpty == false && url[url.startIndex] == "/") {
                    url = convertRelativeURLToBluemixAbsolute(url)
                }

                // method
                let methodString = requestDict.objectForKey("method") as? String
                var method : HttpMethod = HttpMethod.GET;
                switch(methodString){
                    case "GET"?:
                        method = HttpMethod.GET
                        break
                    case "POST"?:
                        method = HttpMethod.POST
                        break
                    case "PUT"?:
                        method = HttpMethod.PUT
                        break
                    case "DELETE"?:
                        method = HttpMethod.DELETE
                        break
                    case "TRACE"?:
                        method = HttpMethod.TRACE
                        break
                    case "HEAD"?:
                        method = HttpMethod.HEAD
                        break
                    case "OPTIONS"?:
                        method = HttpMethod.OPTIONS
                        break
                    case "CONNECT"?:
                        method = HttpMethod.CONNECT
                        break
                    case "PATCH"?:
                        method = HttpMethod.PATCH
                        break
                    default:
                        method = HttpMethod.GET
                }

                // get the query parameters
                let requestQueryParamsDict = requestDict.objectForKey("queryParameters") as! Dictionary<String,String>

                // timeout
                let timeout = (requestDict.objectForKey("timeout") as! Double) / Double(1000)

                // get the headers
                let requestHeaderDict = requestDict.objectForKey("headers") as! Dictionary<String,String>
                let nativeRequest = Request(url: url, method: method, headers: requestHeaderDict, queryParameters: requestQueryParamsDict, timeout: timeout)
                return nativeRequest
            }

    #endif

    #if swift(>=3.0)
    func packResponse(_ response: Response!,error:Error?=nil) throws -> String {
        var jsonResponse:[String: Any] = [:]
        var responseString: String = ""

        if error != nil {
            jsonResponse["errorDescription"] = error!.localizedDescription
        }
        else {
            jsonResponse["errorCode"] = 0
            jsonResponse["errorDescription"] = ""
        }

        if (response == nil)
        {
            jsonResponse["responseText"] = ""
            jsonResponse["headers"] = []
            jsonResponse["status"] = 0
        }
        else {
            let responseText: String = (response.responseText != nil)    ? response.responseText! : ""
            jsonResponse["responseText"] = responseText


            if response.headers != nil {
                jsonResponse["headers"] = response.headers as! [String:Any]
            }
            else {
                jsonResponse["headers"] = [:]
            }

            jsonResponse["status"] = response.statusCode
        }

        let responseData = try JSONSerialization.data(withJSONObject: jsonResponse, options: [])

        responseString =  String(data: responseData, encoding: .utf8)!//try self.stringifyResponse(jsonResponse) as NSString;
        return responseString as String
    }
    #else
    func packResponse(response: Response!,error:NSError?=nil) throws -> String {
        let jsonResponse:NSMutableDictionary = [:]
        var responseString: NSString = ""

        if error != nil {
            jsonResponse.setObject((error!.localizedDescription), forKey: "errorDescription" as NSCopying)
        }
        else {
            jsonResponse.setObject(Int((0)), forKey: "errorCode" as NSCopying)
            jsonResponse.setObject("", forKey: "errorDescription" as NSCopying)
        }

        if (response == nil)
        {
            jsonResponse.setObject("", forKey: "responseText" as NSCopying)
            jsonResponse.setObject([], forKey:"headers" as NSCopying)
            jsonResponse.setObject(Int(0), forKey:"status" as NSCopying)
        }
        else {
            let responseText: String = (response.responseText != nil)    ? response.responseText! : ""
            jsonResponse.setObject(responseText, forKey: "responseText" as NSCopying)

            if response.headers != nil {
                jsonResponse.setObject(response.headers!, forKey:"headers" as NSCopying)
            }
            else {
                jsonResponse.setObject([], forKey:"headers" as NSCopying)
            }

            jsonResponse.setObject(response.statusCode!, forKey:"status" as NSCopying)
        }

        responseString = try self.stringifyResponse(jsonResponse) as NSString;
        return responseString as String
    }

    #endif

    func stringifyResponse(_ value: AnyObject,prettyPrinted:Bool = false) throws -> String {
        #if swift(>=3.0)
            let options = prettyPrinted ? JSONSerialization.WritingOptions.prettyPrinted : JSONSerialization.WritingOptions(rawValue: 0)
            var jsonString : String? = ""
        #else
            let options = prettyPrinted ? NSJSONWritingOptions.PrettyPrinted : NSJSONWritingOptions(rawValue: 0)
            var jsonString : String? = ""
        #endif

        #if swift(>=3.0)
            if JSONSerialization.isValidJSONObject(value) {
                let data = try JSONSerialization.data(withJSONObject: value, options: options)
                jsonString = NSString(data: data, encoding: String.Encoding.utf8.rawValue) as String?
            }
        #else
            if NSJSONSerialization.isValidJSONObject(value) {
                let data = try NSJSONSerialization.dataWithJSONObject(value, options: options)
                jsonString = NSString(data: data, encoding: NSUTF8StringEncoding) as String?
            }
        #endif
        return jsonString!
    }

    func convertRelativeURLToBluemixAbsolute(_ url:String) -> String {
        let client = BMSClient.sharedInstance
        let bluemixAppRoute: String = client.bluemixAppRoute!

        //do Trim first
        #if swift(>=3.0)
            var appRoute = bluemixAppRoute.trimmingCharacters(in: CharacterSet.whitespaces)
        #else
            var appRoute = bluemixAppRoute.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet());
        #endif

        // Remove trailing slashes
        #if swift(>=3.0)
            if(appRoute.characters.last! == "/") {
                appRoute = appRoute.substring(to: appRoute.endIndex)
            }
        #else
            if(appRoute.characters.last! == "/") {
                appRoute = appRoute.substringToIndex(appRoute.endIndex.predecessor())
            }
        #endif
        appRoute = appRoute + url

        return appRoute
    }
}

File: src/android/CDVBMSRequest.java
/*
*     Copyright 2016 IBM Corp.
*     Licensed under the Apache License, Version 2.0 (the "License");
*     you may not use this file except in compliance with the License.
*     You may obtain a copy of the License at
*     http://www.apache.org/licenses/LICENSE-2.0
*     Unless required by applicable law or agreed to in writing, software
*     distributed under the License is distributed on an "AS IS" BASIS,
*     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*     See the License for the specific language governing permissions and
*     limitations under the License.
*/

package com.ibm.mobilefirstplatform.clientsdk.cordovaplugins.core;

import android.content.Context;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import com.ibm.mobilefirstplatform.clientsdk.android.core.api.*;
import com.ibm.mobilefirstplatform.clientsdk.android.logger.api.*;

import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class CDVBMSRequest extends CordovaPlugin {

    private static final Logger bmsLogger = Logger.getLogger("CDVBMSRequest");

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {

        if (action.equals("initialize")) {
            this.send(args, callbackContext);
        }
        else {
            return false;
        }
        return true;
    }

    public void send(final JSONArray args, final CallbackContext callbackContext) {

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    if (!(args.get(0) instanceof JSONObject)) {
                        String message = "Unable to send request. Invalid parameter. Should be JSON object";
                        bmsLogger.debug(message);
                        callbackContext.error(message);
                        return;
                    }
                    JSONObject jsRequest = args.getJSONObject(0);

                    final Context currentContext = cordova.getActivity();
                    final Request request = unpackRequest(jsRequest);
                    final String body = jsRequest.optString("body", "");

                    request.send(currentContext, body, new ResponseListener() {
                        @Override
                        public void onSuccess(Response response) {
                            try {
                                bmsLogger.debug("Successfully sent request");
                                PluginResult result = new PluginResult(PluginResult.Status.OK, packJavaResponseToJSON(response, null));
                                callbackContext.sendPluginResult(result);
                            }
                            catch (JSONException e) {
                                bmsLogger.debug(e.getLocalizedMessage());
                                callbackContext.error(e.getLocalizedMessage());
                            }
                        }

                        @Override
                        public void onFailure(Response response, Throwable throwable, JSONObject jsonObject) {
                            try {
                                bmsLogger.debug("Failed to send request");
                                PluginResult result = new PluginResult(PluginResult.Status.ERROR, packJavaResponseToJSON(response, throwable));
                                callbackContext.sendPluginResult(result);
                            }
                            catch (JSONException e) {
                                bmsLogger.debug(e.getLocalizedMessage());
                                callbackContext.error(e.getLocalizedMessage());
                            }
                        }
                    });
                }
                catch (JSONException e) {
                    bmsLogger.debug(e.getMessage());
                    callbackContext.error(e.getMessage());
                }
            }
        });
    }

    /**
     * Unpacks a JSONObject to create a native Bluemix Request
     *
     * @param requestDict JSON request that will be converted to a native Bluemix Request Object
     * @return The converted Bluemix Request Object
     * @throws JSONException
     */
    private Request unpackRequest(JSONObject requestDict) throws JSONException {

        // URL
        String url = requestDict.getString("url");

        // HTTP method
        String method = requestDict.getString("method");

        // Timeout
        int timeout = requestDict.optInt("timeout", Request.DEFAULT_TIMEOUT);

        Request nativeRequest = new Request(url, method, timeout);

        // Headers
        // TODO: Discuss this at code review
        if (requestDict.has("headers") && !requestDict.isNull("headers")) {

            JSONObject headers = requestDict.getJSONObject("headers");
            Iterator<?> keys = headers.keys();

            while (keys.hasNext()) {
                String key = (String) keys.next();
                nativeRequest.addHeader(key, headers.getString(key));
            }
        }

        // Query parameters
        // TODO: Discuss this at code review
        if (requestDict.has("queryParameters") && !requestDict.isNull("queryParameters")) {

            JSONObject queryParameters = requestDict.getJSONObject("queryParameters");
            Iterator<?> keys = queryParameters.keys();

            while (keys.hasNext()) {
                String key = (String) keys.next();
                nativeRequest.setQueryParameter(key, queryParameters.getString(key));
            }
        }

        return nativeRequest;
    }

    /**
     * Packs a Bluemix Response object into a JSONObject
     *
     * @param response The native Bluemix Response that will be converted to a JSONObject
     * @return The String representation of the JSONObject
     */
    private JSONObject packJavaResponseToJSON(Response response, Throwable t) throws JSONException {

        JSONObject jsonResponse = new JSONObject();

        jsonResponse.put("responseText", response.getResponseText());
        jsonResponse.put("headers", convertHashMapToJSON(response.getHeaders()));
        jsonResponse.put("statusCode", response.getStatus());

        // Pack error
        // TODO: How to get error code?
        if (t != null) {
            //jsonResponse.put("errorCode", );
            jsonResponse.put("errorDescription", t.getLocalizedMessage());
        }

        return jsonResponse;
    }

    /**
     * Converts a HashMap<String, List<String>> to a JSONObject
     *
     * @param originalMap A HashMap that will be converted to a JSONObject
     * @return The converted JSONObject
     */
    private static JSONObject convertHashMapToJSON(Map<String, List<String>> originalMap) throws JSONException {

        JSONObject convertedJSON = new JSONObject();
        Iterator<Map.Entry<String, List<String>>> it = originalMap.entrySet().iterator();

        while (it.hasNext()) {
            Map.Entry<String, List<String>> pair = it.next();
            String key = pair.getKey();
            List<String> headerValues = pair.getValue();
            for (String value : headerValues) {
                convertedJSON.put(key, value);
            }
        }

        return convertedJSON;
    }
}